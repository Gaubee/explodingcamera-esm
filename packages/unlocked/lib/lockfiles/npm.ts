import { join } from "node:path";
import type {
	CommonLock,
	Importer,
	ImporterSnapshots,
	NPMLockfile,
	PackageSnapshots,
	QualifiedDependencyName,
	CommonLockOptions,
	Dependencies,
} from "../types";
import * as utils from "../utils";

type IParse = {
	(directory: false, file: string): Promise<NPMLockfile>;
	(directory: string, file?: undefined): Promise<NPMLockfile>;
};

export const parse: IParse = async (directory, file): Promise<NPMLockfile> => {
	if (directory === false && file === undefined) throw new Error("Either directory or file must be provided");

	return directory ? await import(join(directory, "package-lock.json")) : JSON.parse(file as string);
};

export const toCommonLockfile = async (lockfile: NPMLockfile, options: CommonLockOptions) => {
	let rootPkg = await utils.readPackageJson(options.projectDirectory, options.packageJsonName);

	if (lockfile.lockfileVersion !== 3 && lockfile.lockfileVersion !== 2)
		throw new Error(`Unsupported package-lock.json version: ${lockfile.version}`);

	let packagesToVersions: Record<string, string> = {};
	let importers: ImporterSnapshots = {};
	let packages: PackageSnapshots = {};

	const resolveConcreteVersion = (deps?: Dependencies) =>
		Object.fromEntries(
			Object.entries(deps ?? {}).map(([specifier, version]) => [
				specifier,
				packagesToVersions[`node_modules/${specifier}`] ?? version,
			]),
		);

	// process packages
	for (let [name, dep] of Object.entries(lockfile.packages ?? {})) {
		if (!name.startsWith("node_modules/")) continue;
		if (dep.link && !dep.resolved?.endsWith(".tgz")) continue;
		let actualName = name.slice("node_modules/".length);
		let qualifiedName: QualifiedDependencyName = `/${actualName}/${dep.version}`;

		const meta = await utils.readPackageMetadata(qualifiedName, options.projectDirectory);

		let dependencies = Object.fromEntries(
			Object.entries(dep.dependencies ?? {}).map(([specifier, version]) => [specifier, version]),
		);

		let optionalDependencies = Object.fromEntries(
			Object.entries(dep.optionalDependencies ?? {}).map(([specifier, version]) => [specifier, version]),
		);

		packagesToVersions[name] = dep.version;
		packages[qualifiedName] = {
			dependencies,
			// TODO: npm handles peerDependencies differently, for now we'll just ignore them
			// peerDependencies,
			// optionalPeerDependencies: Object.keys(dep.peerDependenciesMeta ?? []),

			optionalDependencies,

			version: dep.version,
			optional: dep.optional || undefined,
			hasBin: dep.bin !== undefined || undefined,
			engines: dep.engines,
			dev: dep.dev,

			funding: meta.packageJson?.funding,
			authors: meta.authors,
			spdxLicenseId: meta.packageJson?.license,
			packageJsonPath: meta.packageJsonPath,
			licenseFiles: meta.licensePaths,
		};
	}

	// process specifiers
	for (let [name, data] of Object.entries(lockfile.packages ?? {})) {
		if (!data) continue;

		// these have already been handled by the previous loop
		if (name.startsWith("node_modules/")) continue;

		let uniquePackages = utils.unique([
			...Object.entries(data.dependencies ?? {}),
			...Object.entries(data.devDependencies ?? {}),
			...Object.entries(data.optionalDependencies ?? {}),
		]);

		const newData: Importer = {
			specifiers: Object.fromEntries(uniquePackages),
			dependencies: resolveConcreteVersion(data.dependencies),
			devDependencies: resolveConcreteVersion(data.devDependencies),
			optionalDependencies: resolveConcreteVersion(data.optionalDependencies),
		};

		if (name === "") {
			importers["."] = newData;
			continue;
		}

		// this is probably a workspace package / importer
		if (name.includes("/")) {
			importers[name] = newData;
		}
	}

	return <CommonLock>{
		name: lockfile.name ?? rootPkg?.name ?? "unknown",
		version: lockfile.version ?? rootPkg?.version ?? "0.0.0",
		commonLockVersion: 0,
		lockfileVersion: lockfile.lockfileVersion,
		lockfileType: "npm",
		path: options?.projectDirectory ?? undefined,
		importers,
		packages,
	};
};
