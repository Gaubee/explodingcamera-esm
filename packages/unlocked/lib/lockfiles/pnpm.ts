import yaml from "js-yaml";
import {
	findQualifiedDependencyLicenseFile,
	read,
	readPackageJson,
	readQualifiedDependencyPackageJson,
} from "../utils";

import type { PackageJson } from "@npm/types";
import type { Lockfile } from "@pnpm/lockfile-types";
import type { CommonLock, LockDependency, PackageSnapshots, QualifiedDependencyName } from "../types";

type IParse = {
	(directory: false, file: string): Promise<Lockfile>;
	(directory: string, file?: undefined): Promise<Lockfile>;
};

export const parse: IParse = async (directory, file): Promise<Lockfile> => {
	if (directory === false && file === undefined) throw new Error("Either directory or file must be provided");
	const lockfile = directory ? await read(directory, "pnpm-lock.yaml") : (file as string);
	return yaml.load(lockfile) as Lockfile;
};

type ToCommonLockfileOptions = {
	/**
	 * Absolute path to the project root directory
	 */
	projectDirectory?: string;

	/**
	 * If true, dependencies will not be resolved to their package.json
	 * This is useful if you only want to get the lockfile structure
	 * and not the package.json metadata
	 */
	skipResolve?: true;

	packageJsonName?: string;
};

export const toCommonLockfile = async (lockfile: Lockfile, options?: ToCommonLockfileOptions) => {
	let pkg = await readPackageJson(options?.projectDirectory, options?.packageJsonName);

	if (lockfile.packages == null)
		return {
			name: pkg?.name ?? "unknown",
			version: pkg?.version ?? "0.0.0",
			lockfileVersion: 1,
			lockfileType: "pnpm",
		} satisfies CommonLock;

	let packagePromises = Object.entries(lockfile.packages).map(async ([name, dep]) => {
		// resolve the qualified name to a name that can be used to resolve the package
		// e.g. "/is-positive/3.1.0" -> "is-positive"

		let dependencyPkg: PackageJson | undefined;
		let dependencyLicensePath: string | undefined;
		if (!options?.skipResolve && options?.projectDirectory) {
			try {
				dependencyPkg = await readQualifiedDependencyPackageJson(
					name as QualifiedDependencyName,
					options.projectDirectory,
				);
				dependencyLicensePath = await findQualifiedDependencyLicenseFile(
					name as QualifiedDependencyName,
					options.projectDirectory,
				);
			} catch (_) {}
		}

		let dependency = {
			version: dep.version,
			dependencies: dep.dependencies,
			peerDependencies: dep.peerDependencies,
			optionalPeerDependencies: Object.keys(dep.peerDependenciesMeta ?? []),
			optionalDependencies: dep.optionalDependencies,
			optional: dep.optional,
			hasBin: dep.hasBin,
			cpu: dep.cpu,
			os: dep.os,
			engines: dep.engines,
			dev: dep.dev,
			patched: dep.patched,
			requiresBuild: dep.requiresBuild,
			resolution: dep.resolution,

			author: dependencyPkg?.author,
			spdxLicenseId: dependencyPkg?.license,
			licenseFile: dependencyLicensePath,
		} satisfies LockDependency;

		return [name as QualifiedDependencyName, dependency];
	});

	let packages = Object.fromEntries(await Promise.all(packagePromises)) as PackageSnapshots;

	return {
		name: pkg?.name ?? "unknown",
		version: pkg?.version ?? "0.0.0",
		lockfileVersion: 1,
		lockfileType: "pnpm",
		packages,
		importers: lockfile.importers,
	} satisfies CommonLock;
};
