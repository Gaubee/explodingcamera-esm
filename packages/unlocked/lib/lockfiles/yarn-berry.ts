import type { PackageJson } from "@npm/types";
import { parseSyml } from "@yarnpkg/parsers";
import { readFile } from "node:fs/promises";
import { join } from "node:path";

import type { CommonLock, LockDependency } from "../types";
import type { YarnLock } from "./yarn";

type IParse = {
	(directory: false, file: string): Promise<YarnLock<"2">>;
	(directory: string, file?: undefined): Promise<YarnLock<"2">>;
};

export const parse: IParse = async (directory, file): Promise<YarnLock<"2">> => {
	if (directory === false && file === undefined) throw new Error("Either directory or file must be provided");
	const lockfile = directory ? await readFile(join(directory, "yarn.lock"), "utf8") : (file as string);
	const res = parseSyml(lockfile);

	return {
		object: res,
		lockfileType: "2",
	};
};

export const toCommonLockfile = (lockfile: YarnLock<"2">, pkg?: PackageJson) => {
	const dependencies = Object.entries(lockfile.object).reduce(
		(acc, [name, dep]): { [moduleName: string]: LockDependency } => {
			acc[name] = {
				dependencies: dep.dependencies,
				optionalDependencies: dep.optionalDependencies,
				version: dep.version,
			} as LockDependency;
			return acc;
		},
		{} as { [moduleName: string]: LockDependency },
	);

	return {
		name: pkg?.name ?? "unknown",
		version: pkg?.version ?? "0.0.0",
		lockfileVersion: 2,
		// dependencies,
		lockfileType: "yarn-v2",
	} satisfies CommonLock;
};
