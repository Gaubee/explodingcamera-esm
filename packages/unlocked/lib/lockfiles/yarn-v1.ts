import { parse as parseYarn } from "@yarnpkg/lockfile";
import { join } from "node:path";

import type { YarnLock } from "./yarn";
import type { CommonLock, LockDependency, PackageJson } from "../types";
import { readFile } from "node:fs/promises";

// Notes
// - yarn v1 lockfiles are not sorted
// - yarn v1 lockfiles do not have a `dev` field, so for now we will assume that all dependencies are not dev dependencies

const LOCKFILE_PREFIX = `# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1`;

type IParse = {
	(directory: false, file: string): Promise<YarnLock<"1">>;
	(directory: string, file?: undefined): Promise<YarnLock<"1">>;
};

export const parse: IParse = async (directory, file): Promise<YarnLock<"1">> => {
	if (directory === false && file === undefined) throw new Error("Either directory or file must be provided");
	const lockfile = directory ? await readFile(join(directory, "yarn.lock"), "utf8") : (file as string);
	if (lockfile === undefined) throw new Error("No lockfile found");
	if (lockfile === "") throw new Error("Lockfile is empty");
	if (!lockfile.startsWith(LOCKFILE_PREFIX)) throw new Error("Lockfile is not a yarn v1 lockfile");
	const res = parseYarn(lockfile);

	return {
		lockfileType: "1",
		object: res.object,
	};
};

export const toCommonLockfile = (lockfile: YarnLock<"1">, pkg?: PackageJson) => {
	const dependencies = Object.entries(lockfile.object).reduce(
		(acc, [name, dep]): { [moduleName: string]: LockDependency } => {
			acc[name] = {
				dependencies: dep.dependencies,
				optionalDependencies: dep.optionalDependencies,
				version: dep.version,
			} as LockDependency;
			return acc;
		},
		{} as { [moduleName: string]: LockDependency },
	);

	return {
		name: pkg?.name ?? "unknown",
		version: pkg?.version ?? "0.0.0",
		lockfileVersion: 1,
		// dependencies,
		lockfileType: "yarn-v1",
		commonLockVersion: 0,
	} satisfies CommonLock;
};
