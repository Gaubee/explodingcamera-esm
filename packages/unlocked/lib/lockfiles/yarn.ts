import { readFile } from "node:fs/promises";
import { join } from "node:path";

export interface YarnLock<TType extends "1" | "2"> {
	object: YarnLockDeps;
	dependencies?: YarnLockDeps;
	lockfileType: TType;
}

export interface YarnLockDeps {
	[depName: string]: YarnLockDep;
}

export interface YarnLockDep {
	version: string;
	dependencies?: {
		[depName: string]: string;
	};
	optionalDependencies?: {
		[depName2: string]: string;
	};
}

export const detectYarnVersion = async (directory: string) => {
	const lock = await readFile(join(directory, "yarn.lock"), "utf8");
	const match = lock.match(/^# yarn v(\d+)$/m);
	return match ? match[1] : "1";
};
