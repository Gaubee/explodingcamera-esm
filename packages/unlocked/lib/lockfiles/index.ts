import { fileExists } from "../utils";
import { detectYarnVersion } from "./yarn";

export type LockfileType = "yarn-v1" | "yarn-v2" | "pnpm" | "npm";

export type CommonLock = {
	name: string;
	version: string;
	lockfileType: LockfileType;
	lockfileVersion: number;
	packageIntegrity?: string;
	preserveSymlinks?: boolean;
	requires?: boolean;
	dependencies?: { [moduleName: string]: LockDependency };
};

export type LockDependency = {
	version: string;
	integrity?: string;
	resolved?: string;
	bundled?: boolean;
	dev?: boolean;
	optional?: boolean;
	requires?: { [moduleName: string]: string };
	dependencies?: { [moduleName: string]: LockDependency };
};

export const detectLockfileType = async (directory: string): Promise<LockfileType | undefined> => {
	if (await fileExists(directory, "pnpm-lock.yaml")) {
		return "pnpm";
	} else if (await fileExists(directory, "yarn.lock")) {
		const yarnVersion = await detectYarnVersion(directory);
		if (yarnVersion === "1") return "yarn-v1";
		return "yarn-v2";
	} else if (await fileExists(directory, "package-lock.json")) {
		// last priority because it might still exist in a project that uses a different lockfile
		return "npm";
	} else {
		return undefined;
	}
};
