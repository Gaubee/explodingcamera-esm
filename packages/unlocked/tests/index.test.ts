import { join } from "node:path";
import { test, expect } from "vitest";

import { parse as parsePnpm, toCommonLockfile } from "../lib/lockfiles/pnpm";

test("resolve-dependencies", async () => {
	let projectDirectory = join(__dirname, "fixtures", "generic");

	const lockfile = await parsePnpm(projectDirectory);
	expect(lockfile).toMatchSnapshot();

	const commonLockfile = await toCommonLockfile(lockfile, {
		packageJsonName: "package.json",
		projectDirectory,
	});

	expect(commonLockfile).toMatchSnapshot();
});
