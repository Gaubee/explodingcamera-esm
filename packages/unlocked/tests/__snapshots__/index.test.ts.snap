// Vitest Snapshot v1

exports[`resolve-dependencies 1`] = `
{
  "dependencies": {
    "tslib": "2.5.0",
  },
  "lockfileVersion": 5.4,
  "packages": {
    "/tslib/2.5.0": {
      "dev": false,
      "resolution": {
        "integrity": "sha512-336iVw3rtn2BUK7ORdIAHTyxHGRIHVReokCR3XjbckJMK7ms8FysBfhLR8IXnAgy7T0PTPNBWKiH514FOW/WSg==",
      },
    },
  },
  "specifiers": {
    "tslib": "^2.5.0",
  },
}
`;

exports[`resolve-dependencies 2`] = `
{
  "importers": undefined,
  "lockfileType": "pnpm",
  "lockfileVersion": 1,
  "name": "generic",
  "packages": {
    "/tslib/2.5.0": {
      "author": "Microsoft Corp.",
      "cpu": undefined,
      "dependencies": undefined,
      "dev": false,
      "engines": undefined,
      "hasBin": undefined,
      "licenseFile": undefined,
      "optional": undefined,
      "optionalDependencies": undefined,
      "optionalPeerDependencies": [],
      "os": undefined,
      "patched": undefined,
      "peerDependencies": undefined,
      "requiresBuild": undefined,
      "resolution": {
        "integrity": "sha512-336iVw3rtn2BUK7ORdIAHTyxHGRIHVReokCR3XjbckJMK7ms8FysBfhLR8IXnAgy7T0PTPNBWKiH514FOW/WSg==",
      },
      "spdxLicenseId": "0BSD",
      "version": undefined,
    },
  },
  "version": "1.0.0",
}
`;
